### PR TITLE
Fixes handling of role code.

### PR DESCRIPTION
**Pull request by justinlittman**
addresses problem in #1290

## Why was this change made?
To satisfy the requirement:
```
If there's a valueURI without an authority attribute, that's valid and doesn't need to be reported. If the value is three letters, let it pass without the authority attribute and report it as an error. If the value is not three letters, something unpredictable is wrong with the data and it should fail.
```

## How was this change tested?
Unit

## Which documentation and/or configurations were updated?
NA

### CODE REVIEW

File: app/services/cocina/from_fedora/descriptive/contributor.rb [hunk 1]
```diff
@@ -110,7 +110,6 @@ def names
 
         # rubocop:disable Metrics/AbcSize
         # rubocop:disable Metrics/CyclomaticComplexity
-        # rubocop:disable Metrics/PerceivedComplexity
```
Comment (jcoyne) [inline app/services/cocina/from_fedora/descriptive/contributor.rb:113]: 👏

File: app/services/cocina/from_fedora/descriptive/contributor.rb [hunk 2]
```diff
@@ -134,13 +133,23 @@ def roles_for(name)
...
 
         def type_for(type)
           Honeybadger.notify('[DATA ERROR] Contributor type incorrectly capitalized', { tags: 'data_error' }) if type.downcase != type
           ROLES.fetch(type.downcase)
         end
+
+        def check_code(role_code, role_authority)
+          return if role_code.nil? || role_authority
+
+          if role_code&.content.present? && role_code.content.size == 3
```
Comment (jcoyne) [inline app/services/cocina/from_fedora/descriptive/contributor.rb:146]: I don't think you need `&.` here because you've already checked that it's not nil.

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/services/cocina/from_fedora/descriptive/contributor.rb [hunk 3]
```diff
@@ -134,13 +133,23 @@ def roles_for(name)
...
+            Honeybadger.notify('[DATA ERROR] Contributor role code is missing authority', { tags: 'data_error' })
+            return
+          end
+
+          raise Cocina::Mapper::InvalidDescMetadata, "Contributor role code is missing and has unexpected value: #{role_code&.content}"
```
Comment (jcoyne) [inline app/services/cocina/from_fedora/descriptive/contributor.rb:151]: Nil check isn't needed here either.

**Review by jcoyne — CHANGES_REQUESTED** [review]

**Comment by ndushay** [comment]
Please assign #1290 to Arcadia after this is merged.

File: app/services/cocina/from_fedora/descriptive/contributor.rb (hunk 2 above)
Comment (justinlittman) [reply app/services/cocina/from_fedora/descriptive/contributor.rb]: Fixed.

File: app/services/cocina/from_fedora/descriptive/contributor.rb (hunk 3 above)
Comment (justinlittman) [reply app/services/cocina/from_fedora/descriptive/contributor.rb]: Fixed.

File: .rubocop_todo.yml [hunk 4]
```diff
@@ -23,30 +23,30 @@ Lint/SuppressedException:
...
 # Configuration parameters: IgnoredMethods.
 Metrics/AbcSize:
   Max: 38
 
-# Offense count: 4
+# Offense count: 7
 # Configuration parameters: CountComments, CountAsOne.
 Metrics/ClassLength:
-  Max: 119
+  Max: 118
```
Comment (jcoyne) [inline .rubocop_todo.yml:34]: 👏

File: .rubocop_todo.yml [hunk 5]
```diff
@@ -23,30 +23,30 @@ Lint/SuppressedException:
...
 # Configuration parameters: CountComments, CountAsOne, ExcludedMethods.
 Metrics/MethodLength:
   Max: 43
 
-# Offense count: 8
+# Offense count: 16
 # Configuration parameters: IgnoredMethods.
 Metrics/PerceivedComplexity:
-  Max: 13
+  Max: 15
```
Comment (jcoyne) [inline .rubocop_todo.yml:49]: 😢  Can you add an exception on one method rather than increasing the threshold for all methods?

**Review by jcoyne — CHANGES_REQUESTED** [review]

**Review by jcoyne — APPROVED** [review]

**Pull request merged**